src/families/cosmos/app/index.js

```diff
@@ -0,0 +1,298 @@
+          tmp = response.slice(pos, pos + mcuVersionLen - 1);
+        }
+        const mcuVersion = tmp.toString();
+
+        return {
+          return_code: returnCode,
+          error_message: errorCodeToString(returnCode),
+          // //
+          targetId,
+          seVersion,
+          flag,
+          mcuVersion,
+        };
+      }, processErrorResponse);
+  }
+
+  async publicKey(path) {
+    const serializedPath = await this.serializePath(path);
+
+    switch (this.versionResponse.major) {
+      case 1:
+        return publicKeyv1(this, serializedPath);
+      case 2: {
+        const data = Buffer.concat([
+          CosmosApp.serializeHRP("cosmos"),
+          serializedPath,
+        ]);
+        return publicKeyv2(this, data);
+      }
+      default:
+        return {
+          return_code: 0x6400,
+          error_message: "App Version is not supported",
+        };
+    }
+  }
+
+  async getAddressAndPubKey(path, hrp, validateOnDevice) {
+    const serializedPath = await this.serializePath(path);
+    const data = Buffer.concat([CosmosApp.serializeHRP(hrp), serializedPath]);
+    return this.transport
+      .send(CLA, INS.GET_ADDR_SECP256K1, validateOnDevice ? 1 : 0, 0, data, [
+        0x9000,
+      ])
+      .then((response) => {
+        const errorCodeData = response.slice(-2);
+        const returnCode = errorCodeData[0] * 256 + errorCodeData[1];
+
+        const compressedPk = Buffer.from(response.slice(0, 33));
+        const bech32Address = Buffer.from(response.slice(33, -2)).toString();
+
+        return {
+          bech32_address: bech32Address,
+          compressed_pk: compressedPk,
+          return_code: returnCode,
+          error_message: errorCodeToString(returnCode),
+        };
+      }, processErrorResponse);
+  }
+
+  async signSendChunk(chunkIdx, chunkNum, chunk) {
+    switch (this.versionResponse.major) {
+      case 1:
+        return signSendChunkv1(this, chunkIdx, chunkNum, chunk);
+      case 2:
+        return signSendChunkv2(this, chunkIdx, chunkNum, chunk);
+      default:
+        return {
+          return_code: 0x6400,
+          error_message: "App Version is not supported",
+        };
+    }
+  }
+
+  async sign(path, message) {
+    const chunks = await this.signGetChunks(path, message);
+
+    return this.signSendChunk(1, chunks.length, chunks[0], [0x9000]).then(
+      async (response) => {
+        let result = {
+          return_code: response.return_code,
+          error_message: response.error_message,
+          signature: null,
+        };
+
+        for (let i = 1; i < chunks.length; i += 1) {
+          // eslint-disable-next-line no-await-in-loop
+          result = await this.signSendChunk(1 + i, chunks.length, chunks[i]);
+          if (result.return_code !== 0x9000) {
+            break;
+          }
+        }
+
+        return {
+          return_code: result.return_code,
+          error_message: result.error_message,
```

Comment:
the fact there is this "weird" object with error_message makes me thing it's not going to be very well remapped as other coins. we should just "let it throw" like other implementations.

meaning when you refuse on device it's correctly identified as UserRefusedOnDevice and so on. (we based all our work on the fact we have "TransportStatusError".